pipeline_dp/pipeline_operations.py

```diff
@@ -86,6 +90,50 @@ def group_by_key(self, col, stage_name: str):
     def filter(self, col, fn, stage_name: str):
         return col | stage_name >> beam.Filter(fn)
 
+    def filter_by_key(self, col, public_partitions, data_extractors, stage_name: str):
+        """Filter out the partitions that are not not meant to be public.
+
+        Args:
+          col: input collection
+          public_partitions: collection of public partition keys
+          stage_name: name of the stage
+
+        Returns:
+          A filtered collection containing only data belonging to public_partitions
+
+        """
+        class PartitionsFilterJoin(beam.DoFn):
+          def process(self, joined_data):
+            key, rest = joined_data
+            values, is_public = rest.get('values'), rest.get('is_public')
+
+            # TODO the Issue #4 says this is blocked on other tasks. Revisit
+            # this once unblocked
+            if not values:
+              values = [None]
+
+            if is_public:
+              for value in values:
+                yield key, value
+
+        def is_public(col):
+          return col[0] in public_partitions
+
+        VALUES = 0
+        IS_PUBLIC = 1
+
+        if public_partitions is None:
+          raise TypeError("Must provide a valid public_partitions")
+
+        col = col | beam.Map(lambda x: (data_extractors.partition_extractor(x), x))
```

Comment:
Please add stage description for each PTransform applications (i.e. col | "some description" >> beam.Map() )

Explanation: If description is empty, Beam sets default name, something like "Map". And if there >1 non-named beam.Map it leads to name collision.

Reply:
Done.